File: apps/services/gateway/gateway/routes/people/directory.py
"""People Center · the directory and the person page (WS-28b).

Spec: ``project-docs/specs/people_center_app.md`` §3.1, §3.2, §5.2.

Three reads. Every one of them answers the same two questions in the same
order — *may this caller reach the surface at all* (``feature:people``, on the
router) and *may they see the HR half* (``admin:members:read``, projected) —
and neither answer is computed here. Both are imported.
"""

from __future__ import annotations

from typing import Any

from acb_auth import UserContext, get_current_user
from fastapi import Depends, HTTPException
from gateway.routes.people.core import (
    STATUSES,
    _tenant_session,
    can_manage_people,
    can_read_hr_fields,
    has_login,
    router,
)
from gateway.routes.tasks.people import _row_to_person
from pydantic import BaseModel
from sqlalchemy import text

#: A working week, used to turn estimate minutes into the load bar's hours.
#: A constant rather than a setting: the bar is a rough signal, and a tunable
#: nobody tunes is a config surface pretending to be a feature.
MINUTES_PER_HOUR = 60


class DirectoryResponse(BaseModel):
    rows: list[dict[str, Any]]
    total: int
    #: False when the HR half came back nulled. The UI needs this to say
    #: "restricted" rather than "none" — a blank skills strip meaning "you may
    #: not see this" and one meaning "nobody filled it in" are different facts,
    #: and a UI that cannot tell them apart will pick the wrong one (§3.1).
    hr_visible: bool
    #: True when the caller holds `admin:members:manage`. The third of this
    #: app's "say what the answer's shape is" flags, and the one the editor
    #: needs: §3.2 renders write controls **absent** rather than disabled, and
    #: the only alternative to being told is drawing the button and letting the
    #: click discover a 403 — which teaches people to hunt for permissions they
    #: may never get.
    can_manage: bool = False


class WorkResponse(BaseModel):
    rows: list[dict[str, Any]]
    total: int
    #: False when the caller does not hold `feature:projects`. Same distinction
    #: as `hr_visible`: "this surface is not yours" is not "they have no work".
    available: bool


def build_directory_filters(
    *,
    hr: bool,
    q: str = "",
    department: str | None = None,
    team: str | None = None,
    status: str | None = None,
    skill: str | None = None,
    has_capacity: bool = False,
) -> tuple[list[str], dict[str, Any]]:
    """The directory's WHERE clauses and binds — pure, so the rules are testable.

    Extracted from the route deliberately: the interesting decisions here are
    all *whether a clause is included at all*, and that is a question about
    permissions, not about rows. Proving it through a database fake would prove
    it about the fake.

    Three of them are the same rule wearing different hats — **an HR-derived
    filter is dropped entirely without `admin:members:read`**:

    - the ``q`` skills clause, because matching on a column that is then
      stripped from the response turns the search box into an oracle for the
      very field the projection exists to hide;
    - ``skill``, for the same reason, more directly;
    - ``has_capacity``, which is derived from the restricted capacity trio.

    Dropping them silently would be wrong on its own, which is why the response
    carries ``hr_visible``: the UI must be able to say the filter did not apply
    rather than present an unfiltered list as a filtered one.
    """
    clauses: list[str] = ["true"]
    params: dict[str, Any] = {}
    if (q or "").strip():
        match = "(name ILIKE :q OR role ILIKE :q OR title ILIKE :q OR department ILIKE :q"
        if hr:
            match += " OR EXISTS (SELECT 1 FROM unnest(skills) s WHERE s ILIKE :q)"
        clauses.append(match + ")")
        params["q"] = f"%{q.strip()}%"
    if department:
        clauses.append("lower(department) = :department")
        params["department"] = department.strip().lower()
    if team:
        clauses.append("lower(team) = :team")
        params["team"] = team.strip().lower()
    if status:
        clauses.append("status = :status")
        params["status"] = status
    if hr and skill:
        clauses.append("EXISTS (SELECT 1 FROM unnest(skills) s WHERE s ILIKE :skill)")
        params["skill"] = f"%{skill.strip()}%"
    if hr and has_capacity:
        clauses.append("COALESCE(available_hours_per_week, 0) > 0")
    return clauses, params


@router.get("/", response_model=DirectoryResponse)
async def list_directory(
    user: UserContext = Depends(get_current_user),
    q: str = "",
    department: str | None = None,
    team: str | None = None,
    status: str | None = None,
    skill: str | None = None,
    has_capacity: bool = False,
) -> DirectoryResponse:
    """The directory (§3.1).

    ``q`` matches name, title/role and department for everyone — and skills
    **only** for a caller who may see skills. Matching on a column that is then
    stripped from the response turns the search box into an oracle for the very
    field the projection exists to hide; ``/tasks/people`` already made that
    call and this repeats it deliberately rather than diverging.

    ``skill`` and ``has_capacity`` are HR-derived filters, so they are ignored
    outright without ``admin:members:read`` — for the same reason, and stated
    in the response via ``hr_visible`` so the UI does not silently show an
    unfiltered list as though the filter had applied.
    """
    hr = can_read_hr_fields(user)
    if status is not None and status not in STATUSES:
        raise HTTPException(
            status_code=422, detail=f"Unknown status. One of: {list(STATUSES)}.",
        )
    clauses, params = build_directory_filters(
        hr=hr, q=q, department=department, team=team, status=status,
        skill=skill, has_capacity=has_capacity,
    )

    async with _tenant_session() as db:
        rows = (await db.execute(
            text(
                "SELECT * FROM gtd_people WHERE " + " AND ".join(clauses)
                + " ORDER BY department NULLS LAST, name"
            ),
            params,
        )).fetchall()
        people = [_row_to_person(r, include_hr=hr).model_dump() for r in rows]
    return DirectoryResponse(rows=people, total=len(people), hr_visible=hr,
                             can_manage=can_manage_people(user))


@router.get("/facets")
async def list_facets(user: UserContext = Depends(get_current_user)) -> dict:
    """The filter vocabulary, derived from the rows rather than configured.

    A department list somebody has to maintain is a department list that goes
    stale the first time the org changes — the same call ``/projects/my/contexts``
    made for GTD contexts.
    """
    async with _tenant_session() as db:
        rows = (await db.execute(
            text(
                "SELECT department, team, count(*) AS total FROM gtd_people "
                "WHERE department IS NOT NULL GROUP BY department, team "
                "ORDER BY department, team"
            ),
        )).fetchall()
    departments: dict[str, int] = {}
    teams: list[dict[str, Any]] = []
    for row in rows:
        departments[row.department] = departments.get(row.department, 0) + int(row.total)
        if row.team:
            teams.append({"department": row.department, "team": row.team,
                          "total": int(row.total)})
    return {
        "departments": [{"department": d, "total": t} for d, t in departments.items()],
        "teams": teams,
        "statuses": list(STATUSES),
    }


@router.get("/{person_id}")
async def get_person(
    person_id: str, user: UserContext = Depends(get_current_user),
) -> dict:
    """One person (§3.2 panels 1-3).

    404 for a person who does not exist. There is deliberately no 403 anywhere
    in this app: the directory is open to feature holders, and what varies is
    the *shape* of the answer, not whether one comes back.
    """
    hr = can_read_hr_fields(user)
    async with _tenant_session() as db:
        row = (await db.execute(
            text("SELECT * FROM gtd_people WHERE id = CAST(:id AS uuid)"),
            {"id": person_id},
        )).fetchone()
        if row is None:
            raise HTTPException(status_code=404, detail="No such person")
        person = _row_to_person(row, include_hr=hr).model_dump()
        person["has_login"] = await has_login(db, getattr(row, "email", None))
        # An address moved aside by migration 148 because another row claimed
        # it. Surfaced rather than hidden: it means a human still has to decide
        # which row is the real person.
        person["email_conflict"] = getattr(row, "email_conflict", None)
        person["manager"] = await _manager_name(db, getattr(row, "manager_id", None))
        person["load"] = (
            await compute_load(db, getattr(row, "email", None)) if hr else None
        )
    person["hr_visible"] = hr
    # Independent of `hr_visible` on purpose. The two permissions are separate
    # grants, and an admin who may edit a record but not read its HR half is a
    # real principal — the editor has to open with the skills strip restricted
    # and the save button present.
    person["can_manage"] = can_manage_people(user)
    return person


async def _manager_name(db: Any, manager_id: Any) -> str | None:
    if not manager_id:
        return None
    row = (await db.execute(
        text("SELECT name FROM gtd_people WHERE id = CAST(:id AS uuid)"),
        {"id": str(manager_id)},
    )).fetchone()
    return str(row.name) if row is not None else None


async def compute_load(db: Any, email: str | None) -> dict[str, Any]:
    """Load from OPEN ASSIGNED TASKS, not from the typed column (§5.2).

    ``gtd_people.current_load_hours_per_week`` is a number somebody typed once;
    it is stale the moment anyone assigns anything. This counts what the person
    is actually holding.

    The honest part is ``unestimated``: a task with no estimate contributes
    nothing to the hours, so a bar built only from the sum would show somebody
    with thirty un-estimated tasks as completely free. The count travels with
    the number so the UI can say "plus 30 with no estimate" instead of lying.
    """
    if not (email or "").strip():
        return {"open_tasks": 0, "estimated_hours": 0.0, "unestimated": 0}
    row = (await db.execute(
        text(
            """SELECT count(*)                                   AS open_tasks,
                      COALESCE(sum(t.estimate_mins), 0)           AS mins,
                      count(*) FILTER (WHERE t.estimate_mins IS NULL)
                                                                  AS unestimated
                 FROM pm_tasks t
                 JOIN pm_task_statuses s ON s.id = t.status_id
                 JOIN pm_task_assignees a ON a.task_id = t.id
                WHERE lower(a.assignee) = :who
                  AND t.archived_at IS NULL
                  AND s.category NOT IN ('done', 'cancelled')"""
        ),
        {"who": email.strip().lower()},
    )).fetchone()
    mins = int(getattr(row, "mins", 0) or 0)
    return {
        "open_tasks": int(getattr(row, "open_tasks", 0) or 0),
        "estimated_hours": round(mins / MINUTES_PER_HOUR, 1),
        "unestimated": int(getattr(row, "unestimated", 0) or 0),
    }


@router.get("/{person_id}/work", response_model=WorkResponse)
async def get_person_work(
    person_id: str, user: UserContext = Depends(get_current_user),
) -> WorkResponse:
    """This person's open tasks, scoped by the **viewer's** grants (§3.2 panel 4).

    Scoped by the viewer and not the subject, deliberately: a Sales lead looking
    at an Operations colleague should see the work they share, not that
    person's whole life. The Projects grant model already answers exactly that
    question, so it is imported rather than approximated.

    A caller without ``feature:projects`` gets ``available: false`` and an empty
    list rather than a bare empty list — "this surface is not yours" and "they
    have nothing open" must not render identically.
    """
    from gateway.routes.projects.core import resolve_visibility

    if not user.has_permission("feature:projects"):
        return WorkResponse(rows=[], total=0, available=False)

    async with _tenant_session() as db:
        person = (await db.execute(
            text("SELECT email FROM gtd_people WHERE id = CAST(:id AS uuid)"),
            {"id": person_id},
        )).fetchone()
        if person is None:
            raise HTTPException(status_code=404, detail="No such person")
        email = str(getattr(person, "email", "") or "").strip().lower()
        if not email:
            # Directory-only with no address: assignment targets are addresses,
            # so there is nothing to look up rather than nothing assigned.
            return WorkResponse(rows=[], total=0, available=True)

        vis = await resolve_visibility(db, user)
        params: dict[str, Any] = {"who": email}
        clauses = [
            "t.archived_at IS NULL",
            "s.category NOT IN ('done', 'cancelled')",
            "EXISTS (SELECT 1 FROM pm_task_assignees a "
            "WHERE a.task_id = t.id AND lower(a.assignee) = :who)",
        ]
        if not vis.unrestricted:
            clauses.append(vis.project_clause("t.root_project_id"))
            params.update(vis.params)
        rows = (await db.execute(
            text(
                "SELECT t.id, t.title, t.task_number, t.due_at, t.project_id, "
                "       s.name AS status_name, s.category AS status_category, "
                "       p.name AS project_name "
                "  FROM pm_tasks t "
                "  JOIN pm_task_statuses s ON s.id = t.status_id "
                "  LEFT JOIN pm_projects p ON p.id = t.project_id "
                " WHERE " + " AND ".join(clauses)
                + " ORDER BY t.due_at NULLS LAST, t.created_at"
            ),
            params,
        )).fetchall()

    items = [
        {
            "id": str(r.id), "title": r.title, "task_number": r.task_number,
            "due_at": r.due_at.isoformat() if getattr(r, "due_at", None) else None,
            "project_id": str(r.project_id) if r.project_id else None,
            "project_name": getattr(r, "project_name", None),
            "status_name": r.status_name, "status_category": r.status_category,
        }
        for r in rows
    ]
    return WorkResponse(rows=items, total=len(items), available=True)

File: workbench/control_plane/src/app/people/lib/directory.ts
/**
 * People Center · presentation logic for the directory and person page.
 *
 * Spec: `project-docs/specs/people_center_app.md` §3.1, §3.2, §5.2.
 *
 * Pure functions only. The server already decided which rows come back and
 * whether the HR half is filled in; this decides how they read. Keeping the
 * projection server-side and the presentation here is what stops the browser
 * growing a second, weaker answer to "may this person see skills".
 */

import type { PersonDetail, PersonRow } from "./api";

/** How many skills a directory row shows before collapsing into "＋N". */
export const SKILLS_SHOWN = 3;

/**
 * The initials for an avatar fallback.
 *
 * First and last token, so "Vijay Raghav Varada" reads VV rather than VRV —
 * a three-letter monogram in a 24px circle is a smudge.
 */
export function initials(name: string): string {
  const parts = name.trim().split(/\s+/).filter(Boolean);
  if (parts.length === 0) return "?";
  const first = parts[0][0] ?? "";
  const last = parts.length > 1 ? parts[parts.length - 1][0] ?? "" : "";
  return (first + last).toUpperCase();
}

export type SkillsState =
  | { kind: "restricted" }
  | { kind: "empty" }
  | { kind: "some"; shown: string[]; more: number };

/**
 * What the skills strip should render — and the distinction §3.1 insists on.
 *
 * "You may not see this" and "nobody filled it in" are different facts, and a
 * blank strip cannot tell them apart. `hrVisible` comes from the server's own
 * statement about what it projected, never from guessing at a null.
 */
export function skillsState(
  skills: readonly string[] | undefined,
  hrVisible: boolean
): SkillsState {
  if (!hrVisible) return { kind: "restricted" };
  const list = (skills ?? []).filter((s) => (s ?? "").trim().length > 0);
  if (list.length === 0) return { kind: "empty" };
  return {
    kind: "some",
    shown: list.slice(0, SKILLS_SHOWN),
    more: Math.max(0, list.length - SKILLS_SHOWN),
  };
}

/**
 * Where a skill came from: typed by a human, or extracted from a résumé.
 *
 * A skill nobody stated and a parser inferred should not look like a claim the
 * person made (§3.2 panel 2). Unknown provenance reads as `resume` rather than
 * `stated`, because over-claiming on someone's behalf is the worse error.
 */
export function skillOrigin(
  skill: string,
  source: Record<string, string> | undefined
): "stated" | "resume" {
  // ⚠️ The backend's word for "a person typed it" is `manual` — written by
  // create_person and every PATCH since WS-24 — and `observed` (derived from
  // shipped work, WS-28h's vocabulary) is likewise not a parser inference.
  // The first version of this matched only the literal "stated", which no
  // write path has ever produced, so every hand-typed skill in the product
  // rendered as "extracted from a résumé" — over-claiming in the OTHER
  // direction from the one the docstring above worries about. Found while
  // WS-28h made evidence a real column; `directory.test.ts` pins each word.
  const evidence = (source ?? {})[skill];
  return evidence === "resume" || evidence === undefined ? "resume" : "stated";
}

export interface LoadBar {
  /** 0-100, clamped. */
  percent: number;
  label: string;
  /** True when the bar is built on nothing and should be drawn as unknown. */
  unknown: boolean;
}

/**
 * The capacity bar (§3.2 panel 3, §5.2).
 *
 * One bar, not three numbers. Load is the **computed** figure — hours from open
 * assigned tasks — not `current_load_hours_per_week`, which is a number
 * somebody typed once and which is stale the moment anyone assigns anything.
 *
 * The honest part is `unestimated`. A task with no estimate contributes zero
 * hours, so a bar built only from the sum would show somebody holding thirty
 * un-estimated tasks as completely free. When nothing is estimated the bar
 * refuses to draw a percentage at all rather than drawing a confident zero.
 */
export function loadBar(person: Pick<PersonDetail, "capacity_hours_per_week" | "load">): LoadBar {
  const capacity = person.capacity_hours_per_week ?? 0;
  const load = person.load;
  if (!load || (load.open_tasks === 0 && capacity <= 0)) {
    return { percent: 0, label: "No capacity data", unknown: true };
  }
  const hours = load.estimated_hours ?? 0;
  const tail =
    load.unestimated > 0 ? ` · ${load.unestimated} with no estimate` : "";
  if (capacity <= 0) {
    return {
      percent: 0,
      label: `${load.open_tasks} open${tail}`,
      unknown: true,
    };
  }
  if (hours === 0 && load.unestimated > 0) {
    // Everything they hold is un-estimated: a 0% bar would read as "free".
    return {
      percent: 0,
      label: `${load.open_tasks} open, none estimated`,
      unknown: true,
    };
  }
  return {
    percent: Math.min(100, Math.round((hours / capacity) * 100)),
    label: `${hours}h of ${capacity}h · ${load.open_tasks} open${tail}`,
    unknown: false,
  };
}

/**
 * Group directory rows by department for the list's section headers.
 *
 * People with no department land in one trailing group rather than being
 * dropped — a directory that quietly omits somebody is worse than one with an
 * untidy last section.
 */
export const NO_DEPARTMENT = "Unassigned";

export function groupByDepartment(
  rows: readonly PersonRow[]
): Array<{ department: string; people: PersonRow[] }> {
  const groups = new Map<string, PersonRow[]>();
  for (const row of rows) {
    const key = (row.department ?? "").trim() || NO_DEPARTMENT;
    const bucket = groups.get(key);
    if (bucket) bucket.push(row);
    else groups.set(key, [row]);
  }
  const named = [...groups.entries()]
    .filter(([d]) => d !== NO_DEPARTMENT)
    .sort((a, b) => a[0].localeCompare(b[0]));
  const rest = groups.get(NO_DEPARTMENT);
  return [
    ...named.map(([department, people]) => ({ department, people })),
    ...(rest ? [{ department: NO_DEPARTMENT, people: rest }] : []),
  ];
}

/** The pill styling class for a status, so the vocabulary lives in one place. */
export function statusTone(status: string): "active" | "muted" | "warn" {
  switch (status) {
    case "active":
      return "active";
    case "contractor":
    case "invited":
      return "warn";
    default:
      return "muted";
  }
}
